rosidl_cmake/cmake/rosidl_generate_interfaces.cmake

```diff
@@ -220,6 +220,7 @@ macro(rosidl_generate_interfaces target)
   )
 
   if(NOT _ARG_SKIP_INSTALL)
+    _ament_cmake_export_targets_register_package_hook()
```

Comment:
why do you call this manually?
That seems to be pretty odd.

Reply:
It is odd, and in retrospect I should've added a big comment right next to it.

In `ament_cmake`, hook macros conventionally register `ament_package` extensions. That's the case for [`_ament_cmake_export_targets_register_package_hook()`](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/ament_cmake_export_targets-extras.cmake#L19-L27). Typically, this extension is what actually performs the intended action e.g. `ament_export_targets()` collects them, but it is the [extension](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/cmake/ament_cmake_export_targets_package_hook.cmake) that actually installs the export set.

One normally does not invoke them directly -- [the user-facing macro does](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/cmake/ament_export_targets.cmake#L38). The big issue with it is that you lose control of _when_ the extension is registered, and extensions are invoked in registration order. That order is reflected in the `${PROJECT_NAME}_CONFIG_EXTRAS` variable, which is the same as the order in which the extra CMake files are included by downstream code.

Without this change, it is perfectly possible for [`rosidl_cmake_export_typesupport_targets-extras.cmake`](https://github.com/ros2/rosidl/blob/master/rosidl_cmake/cmake/rosidl_cmake_export_typesupport_targets-extras.cmake.in) to be loaded before [`ament_export_targets-extras.cmake`](https://github.com/ament/ament_cmake/blob/84719051ef2b26070de484e60b8f060ae7a70b06/ament_cmake_export_targets/cmake/ament_cmake_export_targets-extras.cmake.in) and fail. By invoking the hook earlier, I ensure targets are imported first.

Arguably, you can achieve the same by making sure that macros are invoked in an specific order in all generators (not the case right now) but it is fragile IMHO.

Reply:
> Without this change, it is perfectly possible for rosidl_cmake_export_typesupport_targets-extras.cmake to be included before ament_export_targets-extras.cmake and fail to collect any targets (which is what happened to me). By invoking the hook earlier, we ensure targets are imported first.

this is unfortunate, but I don't see a simple solutions ....
Please add a big comment :smiley:

Reply:
See https://github.com/ros2/rosidl/pull/601/commits/a10f1c8288525ad38809300ff9c1d8f0c9eb7a13.

Reply:
danke